compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -1170,6 +1171,52 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             _ => return,
         }
     }
+
+    fn note_fn_method_def_due_to_call(
+        &self,
+        err: &mut Diagnostic,
+        expr: &hir::Expr<'_>,
+        expected: Ty<'_>,
+    ) {
+        let (def_id, ident, callee_str) = if let hir::ExprKind::Call(fun, _) = expr.kind
+            && let hir::ExprKind::Path(hir::QPath::Resolved(_, path)) = fun.kind
+            && let hir::def::Res::Def(def_kind, def_id) = path.res
+            && !matches!(def_kind, hir::def::DefKind::Ctor(..))
+        {
+            (def_id, path.segments[0].ident, "function")
+        } else if let hir::ExprKind::MethodCall(method, ..) = expr.kind
+            && let Some(def_id) = self.typeck_results.borrow().type_dependent_def_id(expr.hir_id)
+            && !matches!(self.tcx.def_kind(def_id), hir::def::DefKind::Ctor(..))
+        {
+            (def_id, method.ident, "method")
+        } else {
+            return;
+        };
+        err.span_note(
+            self.tcx.def_span(def_id),
+            format!("the {callee_str} {ident} is defined here"),
+        );
+
+        if let Some(local_did) = def_id.as_local()
+            && let Some(node) = self.tcx.opt_hir_node(self.tcx.local_def_id_to_hir_id(local_did))
+            && let hir::Node::ImplItem(hir::ImplItem {
+                kind: hir::ImplItemKind::Fn(sig, ..), ..
+            })
+            | hir::Node::Item(hir::Item { kind: hir::ItemKind::Fn(sig, ..), .. }) = node
+            && let ret_span = sig.decl.output.span()
+            && !ret_span.from_expansion()
```

Comment:
Why not from expansions?

Reply:
Mainly to avoid giving weird suggestions in desugaring like `async fn`, though if you think we should still give suggestions in macros, we could check for `desugaring_kind().is_none()` instead.